GH-36953: [MATLAB] Add gateway `arrow.array` function to create Arrow Arrays from MATLAB data

sgilmore10 opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

As discussed in #36855, we think it would be better to move the recommended APIs for the MATLAB Interface directly under the top-level `arrow.*` package. This should help simplify the interface, and will make it easier for users to switch between multiple language bindings. We have already moved the `type` convenience constructors to the `arrow` package.  Now we want to add a gateway function that creates arrays to mirror `PyArrow`. As part of this change, we will modify the array constructors to accept `libmexclass.proxy.Proxy` objects - similar to how the `arrow.type.<Type>` constructors accept  `libmexclass.proxy.Proxy` objects.

### What changes are included in this PR?

1. Added `arrow.array()` gateway function that can be used to construct arrays:

```matlab
>> arrowArray = arrow.array([1 2 3 4]);
>> class(arrowArray)

ans =

    'arrow.array.Float64Array'

>> arrowArray = arrow.array(["A" "B" "C"]);
>> class(arrowArray)

ans =

    'arrow.array.StringArray'

```

2. Added a static `fromMATLAB()` method to all  subclasses of`arrow.array.Array`. 

```matlab
>> array = arrow.array.StringArray.fromMATLAB(["A" "B" "C"])

array = 

[
  "A",
  "B",
  "C"
]

>> array = arrow.array.TimestampArray.fromMATLAB(datetime(2023, 8, 1))

array = 

[
  2023-08-01 00:00:00.000000
]

```

As part of this change, users can no longer use the `arrow.array.Array` subclass constructors to create arrays. Instead, they can use either `arrow.array()` or the static `fromMATLAB` method.


### Are these changes tested?

Updated the existing tests to account for the API changes and added the following new test classes:

1. arrow/internal/validate/tType.m
2. arrow/internal/validate/tShape.m
3. arrow/internal/validate/tRealNumeric.m
4. arrow/internal/validate/tNonsparse.m
5. arrow/internal/validate/tNumeric.m
6. arrow/array/tArray.m

### Are there any user-facing changes?

Yes, we changed the signature of all `arrow.array.Array` subclasses to accept scalar `libmexclass.proxy.Proxy` classes. NOTE: The MATLAB interface is still under active development. 

### Future Directions

1. In a followup PR, we plan on adding a new name-value pair to `arrow.array()` called `Type`, which can be set to an `arrow.type.Type` object. This will let users specify what kind of arrow array they would like to create from MATLAB data.


<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* Closes: #36953

kou reviewed (approved):
+1

(This is too large to review...)

sgilmore10 commented:
Hi @kou, I apologize for the size of this review. I'll make sure to keep my future PRs as small as possible.

I thought about splitting these changes across multiple PRs, but couldn't find a way to do so without mass test failures. 

Again, I apologize for the size of the PR. If it helps, I've summarized the major changes in this PR below:

- Added a new gateway function called `arrow.array()`. This mirrors the design of PyArrow. 
- Added a static `fromMATLAB` method to all concrete subclasses of `arrow.array.Array`. Clients can use these methods if they prefer not to rely on the gateway function.
- We no longer expect or want clients to call the constructors of `arrow.array.Array` subclasses directly. This also mirrors the design of PyArrow.

Best,
Sarah

kou commented:
Thanks for the summary.
(I should have written as "I could understand what you did but this is too large to review...". Sorry.)

We should wait for a review from @kevingurney .

kevingurney commented on matlab/src/matlab/+arrow/+array/Array.m:
```diff
@@ -31,8 +31,11 @@
-        function obj = Array(varargin)
-            obj.Proxy = libmexclass.proxy.Proxy(varargin{:}); 
+        function obj = Array(proxy)
+        arguments
```
Indent?

sgilmore10 replied:
fixed.

kevingurney commented on matlab/src/matlab/+arrow/+array/TimestampArray.m:
```diff
@@ -70,4 +62,26 @@
+            arrow.internal.validate.shape(data);
+
+            validElements = arrow.internal.validate.parseValidElements(data, opts);
+            ptime = arrow.array.TimestampArray.convertToEpochTime(data, opts.TimeUnit);
```
Maybe we should call this variable `epochTime`?

sgilmore10 replied:
done.

kevingurney commented on matlab/src/matlab/+arrow/array.m:
```diff
@@ -0,0 +1,67 @@
+        opts.Valid
+    end
+
+    data = convertCellstrToStrings(data);
```
Maybe we should rename this to `convertCellstrToString`?

sgilmore10 replied:
I agree.

kevingurney commented on matlab/src/matlab/+arrow/array.m:
```diff
@@ -0,0 +1,67 @@
+    switch (classname)
+        case "logical"
+            arrowArray = arrow.array.BooleanArray.fromMATLAB(data, args{:});
+        case "int8"
```
This is a minor comment, but in a lot of other places we have been ordering this list of types as `uint8`, `uint16`, ... rather than "interleaving" the signed and unsigned integer types.

sgilmore10 replied:
That's a good point. I'll reorder the cases.

kevingurney commented on matlab/test/arrow/array/hNumericArray.m:
```diff
@@ -92,20 +92,21 @@ function MaxValueTest(tc)
         end
 
-        function ErrorIfEmptyArrayIsNotTwoDimensional(tc)
+        function AllowNDEmptyArray(tc)
```
`ND` -> `NDimensional`?

sgilmore10 replied:
done.

kevingurney commented on matlab/test/arrow/array/tArray.m:
```diff
@@ -0,0 +1,88 @@
+    methods(Test)        
+        function ArrowArrayOutputType(testCase, MATLABDataArrayTypePair)
+        % Verify arrow.array returns the expected arrow.array.<Type>Array
+        % with respect to the input data array.
```
"...with respect to the input data array's MATLAB class type."

sgilmore10 replied:
done.

kevingurney commented on matlab/test/arrow/array/tArray.m:
```diff
@@ -0,0 +1,88 @@
+        end
+
+        function InferNullsTrue(testCase)
+        % Verify InferNulls is true by default.
```
I think this should say something like "Verify that Valid contains the expected logical vector when InferNulls=true."

kevingurney commented on matlab/test/arrow/array/tArray.m:
```diff
@@ -0,0 +1,88 @@
+        end
+
+        function ValidNameValuePair(testCase)
+        % Verify the Valid property of the arrow array when the Valid
```
"Verify the Valid property returns the expected logical vector..."

sgilmore10 replied:
dibe,

kevingurney commented on matlab/test/arrow/array/tBooleanArray.m:
```diff
@@ -136,19 +136,20 @@ function ErrorIfNonVector(tc)
-            data = tc.MatlabArrayFcn(reshape(logical.empty(0, 0), [1 0 0]));
-            fcn = @() tc.ArrowArrayConstructor(data);
-            tc.verifyError(fcn, "MATLAB:expected2D");
+        function AllowNDEmptyArray(tc)
```
ND -> NDimensional?

sgilmore10 replied:
done.

kevingurney commented on matlab/test/arrow/array/tStringArray.m:
```diff
@@ -136,13 +136,14 @@ function ErrorIfNonVector(tc)
         end
 
-        function ErrorIfEmptyArrayIsNotTwoDimensional(tc)
+        function AllowNDEmptyArray(tc)
```
ND -> NDimensional?

sgilmore10 replied:
done.

kevingurney commented on matlab/test/arrow/array/tTimestampArray.m:
```diff
@@ -134,47 +133,54 @@ function TestInferNulls(testCase, TimeUnit, TimeZone)
+            testCase.verifyEqual(arrowArray.Valid, logical.empty(0, 1));
+            testCase.verifyEqual(toMATLAB(arrowArray), datetime.empty(0, 1));
+
+            % test with ND empty array
```
ND -> NDimensional?

kevingurney commented on matlab/test/arrow/internal/validate/tRealNumeric.m:
```diff
@@ -0,0 +1,47 @@
+
+classdef tRealNumeric < matlab.unittest.TestCase
+    
+    methods(TestClassSetup)
```
I think this can be deleted.

kevingurney commented on matlab/test/arrow/internal/validate/tRealNumeric.m:
```diff
@@ -0,0 +1,47 @@
+    end
+
+    properties(TestParameter)
+        NumericType = struct(int8=@int8,...
```
Could we change this to not be "interleaved"?

kevingurney commented on matlab/test/arrow/internal/validate/tShape.m:
```diff
@@ -0,0 +1,69 @@
+%TREALNUMERIC Unit tests for arrow.internal.validate.shape.
```
TREALNUMERIC -> TSHAPE

sgilmore10 replied:
oops. Thanks for catching this.

kevingurney commented on matlab/test/arrow/internal/validate/tShape.m:
```diff
@@ -0,0 +1,69 @@
+classdef tShape < matlab.unittest.TestCase
+    
+    methods(Test)
+        function ErrorIf2DMatrix(testCase)
```
2D -> 2Dimensional?

sgilmore10 replied:
done.

kevingurney reviewed (changes requested):
Thanks for working on this, @sgilmore10!

Overall, the changes look good to me - just some minor comments.

sgilmore10 commented:
> Thanks for the summary. (I should have written as "I could understand what you did but this is too large to review...". Sorry.)
> 
> We should wait for a review from @kevingurney .

Hi @Kou,

 Looking back, I realize I could have held off on making the changes to the validation functions. That would have reduced the size of this review. Going forward I'll make sure keep my PRs as small as possible. I appreciate the feedback!

Best,
Sarah

sgilmore10 commented on matlab/test/arrow/array/tBooleanArray.m:
```diff
@@ -18,7 +18,7 @@
       properties
         ArrowArrayClassName(1, 1) string = "arrow.array.BooleanArray"
-        ArrowArrayConstructor = @arrow.array.BooleanArray
+        ArrowArrayConstructor = @arrow.array.BooleanArray.fromMATLAB
```
good point. done.

kevingurney commented:
+1
